### ISSUE TITLE
Specify bfcache behavior with proper tests

### ISSUE THREAD

**Issue by saschanaz** [WICG/web-locks#81]
I intend to add a bfcache test where any active lock/request prevents caching, matching the current Chrome Dev 96 behavior.

Is this something you want to change or are you okay with the current Chrome behavior? To me it sounds reasonable enough, and in that case I think we should at least add a note about it, if it's not something can be specced.

**Comment by saschanaz** [comment]
From https://github.com/WICG/web-locks/issues/78#issuecomment-920543922

>Coincidently, the following guidance just got merged into the TAG design guidelines:
https://w3ctag.github.io/design-principles/#support-non-fully-active

It seems this also covers bfcache which should help us spec the behavior.

**Comment by hiroshige-g** [comment]
Great. FYI There are also some docs about writing bfcache/navigation tests at

- https://github.com/web-platform-tests/wpt/blob/master/html/browsers/browsing-the-web/back-forward-cache/README.md
- https://github.com/web-platform-tests/wpt/blob/master/common/dispatcher/README.md

and some more test examples under codereview (they will be refactored before landing though):

- https://chromium-review.googlesource.com/c/chromium/src/+/3198456
- https://chromium-review.googlesource.com/c/chromium/src/+/3198457
- https://chromium-review.googlesource.com/c/chromium/src/+/3201011
- https://chromium-review.googlesource.com/c/chromium/src/+/3222281

**Comment by saschanaz** [comment]
Another FYI, WPT now has some bfcache tests specifically for Web Locks, e.g. https://wpt.live/web-locks/bfcache/held.tentative.https.html

**Comment by rakina** [comment]
It seems like we have a WPT already. Is there anything that needs to be updated on the spec side for this? BTW there's a [guide](https://w3ctag.github.io/design-principles/#support-non-fully-active) on how to handle non-fully active documents (which includes BFCached documents).

**Comment by saschanaz** [comment]
> Is there anything that needs to be updated on the spec side for this?

It currently says nothing about blocking bfcache, so maybe we should add some?

It seems setting [`salvagable`](https://html.spec.whatwg.org/multipage//browsing-the-web.html#concept-document-salvageable) to `false` does that work, but in this way we can't safely set it true again.

@annevk, is `salvagable` the right tool to use here?

**Comment by annevk** [comment]
Yeah, you'd set salvageable to false during the [unloading document cleanup steps](https://html.spec.whatwg.org/#unloading-document-cleanup-steps).

cc @domenic

**Comment by domfarolino** [comment]
This issue has been dormant for a few years, but I'm arriving here because of https://github.com/whatwg/html/pull/11600#issuecomment-4014058921 and https://github.com/whatwg/html/issues/11520.

My understanding is that Chrome's current behavior (or soon-to-be-shipped behavior?) is allowing Documents that are holding web locks to enter the bf-cache without releasing locks. I think this breaks https://w3c.github.io/web-locks/#ref-for-unloading-document-cleanup-steps and https://html.spec.whatwg.org/#blocking-weblock, so I've asked @yoshisatoyanagisawa and @llannasatoll to take a look at this, since they're pretty familiar with this space these days.

**Comment by llannasatoll** [comment]
Yeah Chrome recently updated its behavior: pages holding Web Locks can now enter the bfcache. However, if another active document tries to acquire or query a lock held by a bfcached page, that page is evicted from the cache to release the lock.

Any concerns from other browsers? If not, I will start working on a PR to update the spec.

**Comment by saschanaz** [comment]
Updating the spec wound be too early (only one impl) but certainly an interesting approach.

**Comment by annevk** [comment]
I don't think there's a requirement that there are two implementations before the specification can change? Related, https://www.w3.org/2026/01/webappswg-charter-2026.html#working-mode suggests there should be a PR template, but I'm not seeing it.

**Comment by saschanaz** [comment]
Well two supports, yeah. Support from moz can come if there's some strong proof that this is web compatible.

**Comment by yoshisatoyanagisawa** [comment]
Since it is not in the spec yet, it should be under tentative or so for now. I think there are some tentative WPTs for this added/updated as a part of Chromium change (https://crrev.com/c/6914636). Shall we use those to codify the expected behavior while we discuss the spec update?

**Comment by llannasatoll** [comment]
This behavior just shipped in Chrome M146, so we are currently monitoring it. I will share the data and any findings with you once we have gathered enough data.

The (tentative) [WPTs](https://wpt.fyi/results/web-locks/bfcache?label=experimental&label=master&aligned) we added should be a good reference for the behavior:

- `held.tentative.https.html`: verifies that holding a lock doesn't block bfcache.
- `contention.https.window.html`: verifies the eviction logic upon lock contention.
